File: runtime/java/titan-ast-runtime/src/titan/ast/runtime/Ast.java
package titan.ast.runtime;

import java.util.ArrayList;

/**
 * 抽象语法树.
 *
 * @author tian wei jun
 */
public abstract class Ast {
  public final AstGrammar grammar;

  public ArrayList<Ast> children = new ArrayList<>();

  public Ast(AstGrammar grammar) {
    this.grammar = grammar;
  }
}

File: runtime/cpp/titan-ast-runtime-lib/Ast.cpp
//
// Created by tian wei jun on 2022/11/17 0017.
//

#include "Ast.h"

#include "StringUtils.h"
#include <utility>

Ast::Ast(AstGrammar grammar)
    : grammar(std::move(grammar)), children(std::vector<Ast *>{}) {}

Ast::~Ast() {
  for (auto child : children) {
    delete child;
  }
}

TerminalAst::TerminalAst(AstGrammar grammar, AstToken token)
    : Ast(grammar), token(token) {}

std::string TerminalAst::toString() const { return token.text; }

NonterminalAst::NonterminalAst(AstGrammar grammar, std::string alias)
    : Ast(grammar), alias(alias) {}

std::string NonterminalAst::toString() const {
  std::string displayString;
  if (StringUtils::isNotBlank(&alias)) {
    displayString = grammar.name + "[" + alias + "]";
  } else {
    displayString = grammar.name;
  }
  return displayString;
}
